File: src/kaya_go/scoring.py
"""领地/记分计算 —— 对应 packages/ui/src/services/scoring.ts。

核心逻辑逐函数移植:
  - calculateTerritory      flood fill 确定封口领地
  - floodFill / determineOwner  区域性连通与归属判定
  - calculateEstimatedTerritory  flood fill 兜底 + probabilityMap 补官子
  - countDeadStones         按子色统计死子数

符号约定与 Kaya 一致:Sign -1=白 / 0=空 / 1=黑;
territoryMap 沿用同一约定(-1=白地 / 0=单官中立 / 1=黑地)。
"""
from __future__ import annotations

# 领地归属阈值 —— 与 TS TERRITORY_THRESHOLD=0.2 一致:
# |prob| 低于该值视为接近中性,概率图不强行归属(交给 flood fill 判为单官)。
TERRITORY_THRESHOLD = 0.2


def flood_fill(board: list[list[int]], start: tuple[int, int], visited: set[tuple[int, int]]) -> list[tuple[int, int]]:
    """找与 start 连通的空点区域(栈式 BFS/DFS)。board: signMap(死子已清零)。"""
    height = len(board)
    width = len(board[0]) if board else 0
    region: list[tuple[int, int]] = []
    stack: list[tuple[int, int]] = [start]

    while stack:
        x, y = stack.pop()
        if (x, y) in visited:
            continue
        if x < 0 or x >= width or y < 0 or y >= height:
            continue
        if board[y][x] != 0:
            continue
        visited.add((x, y))
        region.append((x, y))
        stack.append((x + 1, y))
        stack.append((x - 1, y))
        stack.append((x, y + 1))
        stack.append((x, y - 1))

    return region


def determine_owner(board: list[list[int]], region: list[tuple[int, int]]) -> int:
    """判定空域归属:只贴黑→1,只贴白→-1,两边都贴(单官)→0。"""
    height = len(board)
    width = len(board[0]) if board else 0

    touches_black = False
    touches_white = False

    for x, y in region:
        for nx, ny in ((x + 1, y), (x - 1, y), (x, y + 1), (x, y - 1)):
            if nx < 0 or nx >= width or ny < 0 or ny >= height:
                continue
            sign = board[ny][nx]
            if sign == 1:
                touches_black = True
            elif sign == -1:
                touches_white = True

    if touches_black and not touches_white:
        return 1
    if touches_white and not touches_black:
        return -1
    return 0  # Dame (neutral)


def compute_territory(sign_map: list[list[int]], dead_stones: set[tuple[int, int]]) -> dict:
    """精确路径:flood fill 找封口领地,死子先清零。对应 TS calculateTerritory。

    返回 {"blackTerritory": int, "whiteTerritory": int, "territories": list[list[int]]}。
    """
    height = len(sign_map)
    width = len(sign_map[0]) if sign_map else 0

    # 死子清零副本
    modified_board = [row[:] for row in sign_map]
    for x, y in dead_stones:
        if 0 <= y < height and 0 <= x < width:
            modified_board[y][x] = 0

    visited: set[tuple[int, int]] = set()
    territories = [[0] * width for _ in range(height)]
    black_territory = 0
    white_territory = 0

    for y in range(height):
        for x in range(width):
            if modified_board[y][x] == 0 and (x, y) not in visited:
                region = flood_fill(modified_board, (x, y), visited)
                owner = determine_owner(modified_board, region)
                for vx, vy in region:
                    territories[vy][vx] = owner
                if owner == 1:
                    black_territory += len(region)
                elif owner == -1:
                    white_territory += len(region)

    return {"blackTerritory": black_territory, "whiteTerritory": white_territory, "territories": territories}


def compute_estimated_territory(
    sign_map: list[list[int]],
    probability_map: list[list[float]],
    dead_stones: set[tuple[int, int]],
) -> dict:
    """估计路径:flood fill 兜底 + probabilityMap 补官子。对应 TS calculateEstimatedTerritory。

    先算能确定的封口领地;对 flood fill 留下的中性/单官空点,再用概率图判定:
    prob >  0.2 → 黑地;  prob < -0.2 → 白地。
    """
    height = len(sign_map)
    width = len(sign_map[0]) if sign_map else 0

    flood = compute_territory(sign_map, dead_stones)
    territories = flood["territories"]
    black_territory = flood["blackTerritory"]
    white_territory = flood["whiteTerritory"]

    for y in range(height):
        for x in range(width):
            if sign_map[y][x] != 0 or territories[y][x] != 0:
                continue
            row = probability_map[y] if y < len(probability_map) else None
            prob = row[x] if row is not None and x < len(row) else 0.0
            if prob > TERRITORY_THRESHOLD:
                territories[y][x] = 1
                black_territory += 1
            elif prob < -TERRITORY_THRESHOLD:
                territories[y][x] = -1
                white_territory += 1

    return {"blackTerritory": black_territory, "whiteTerritory": white_territory, "territories": territories}


def parse_dead_stones(dead_stones: list[dict]) -> set[tuple[int, int]]:
    """把死子坐标列表(来自 /api/v1/deadstones 的 [{x,y},...])转为 set[(x,y)]。"""
    return {(d["x"], d["y"]) for d in dead_stones}

File: sample_front/src/lib/scoring.ts
/**
 * 领地/记分计算 —— 从前端主应用 `packages/ui/src/services/scoring.ts` 移植。
 *
 * 纯函数、无外部依赖，直接在本项目（sample_front）内运行，无需再请求后端
 * `/api/v1/score`。输入三件套（signMap + deadStones [+ probabilityMap]）已
 * 由前端本地持有，因此记分在 19×19 盘面下为亚毫秒级，同步执行、零网络往返。
 *
 * 符号约定与后端 /api/v1/score 一致：Sign 1=黑 / -1=白 / 0=空；
 * territory 沿用同一约定（-1=白地 / 0=单官中立 / 1=黑地）。
 */

/** 领地归属阈值 —— 与后端 scoring.TERRITORY_THRESHOLD=0.2 一致。 */
export const TERRITORY_THRESHOLD = 0.2;

/** 死子判定阈值 —— 与后端 deadstones.DEAD_STONE_THRESHOLD=0.4 一致。 */
export const DEAD_STONE_THRESHOLD = 0.4;

/** 坐标 key：与后端 deadStones 逐点一致；Set<string> 便于 O(1) 查重。 */
const key = (x: number, y: number): string => `${x},${y}`;

export interface TerritoryResult {
  blackTerritory: number;
  whiteTerritory: number;
  /** -1 = 白地, 0 = 中立/单官, 1 = 黑地 */
  territories: number[][];
}

/**
 * 精确路径：flood fill 找封口领地，死子先清零。对应后端 compute_territory。
 */
export function calculateTerritory(
  signMap: number[][],
  deadStones: Set<string>
): TerritoryResult {
  const height = signMap.length;
  const width = signMap[0]?.length || 0;

  // 死子清零副本
  const modifiedBoard: number[][] = signMap.map((row, y) =>
    row.map((sign, x) => (deadStones.has(key(x, y)) ? 0 : sign))
  );

  const visited = new Set<string>();
  const territories: number[][] = Array.from({ length: height }, () =>
    Array<number>(width).fill(0)
  );

  let blackTerritory = 0;
  let whiteTerritory = 0;

  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      if (modifiedBoard[y][x] !== 0 || visited.has(key(x, y))) continue;

      const region = floodFill(modifiedBoard, [x, y], visited);
      const owner = determineOwner(modifiedBoard, region);

      for (const [vx, vy] of region) territories[vy][vx] = owner;
      if (owner === 1) blackTerritory += region.length;
      else if (owner === -1) whiteTerritory += region.length;
    }
  }

  return { blackTerritory, whiteTerritory, territories };
}

/**
 * 找与 start 连通的空点区域（栈式 DFS）。对应后端 flood_fill。
 */
function floodFill(
  board: number[][],
  start: [number, number],
  visited: Set<string>
): [number, number][] {
  const height = board.length;
  const width = board[0]?.length || 0;
  const region: [number, number][] = [];
  const stack: [number, number][] = [start];

  while (stack.length > 0) {
    const [x, y] = stack.pop()!;
    if (visited.has(key(x, y))) continue;
    if (x < 0 || x >= width || y < 0 || y >= height) continue;
    if (board[y][x] !== 0) continue;

    visited.add(key(x, y));
    region.push([x, y]);
    stack.push([x + 1, y], [x - 1, y], [x, y + 1], [x, y - 1]);
  }

  return region;
}

/**
 * 判定空域归属：只贴黑→1，只贴白→-1，两边都贴（单官）→0。对应后端 determine_owner。
 */
function determineOwner(board: number[][], region: [number, number][]): number {
  const height = board.length;
  const width = board[0]?.length || 0;

  let touchesBlack = false;
  let touchesWhite = false;

  for (const [x, y] of region) {
    if (x + 1 < width && board[y][x + 1] === 1) touchesBlack = true;
    if (x + 1 < width && board[y][x + 1] === -1) touchesWhite = true;
    if (x - 1 >= 0 && board[y][x - 1] === 1) touchesBlack = true;
    if (x - 1 >= 0 && board[y][x - 1] === -1) touchesWhite = true;
    if (y + 1 < height && board[y + 1][x] === 1) touchesBlack = true;
    if (y + 1 < height && board[y + 1][x] === -1) touchesWhite = true;
    if (y - 1 >= 0 && board[y - 1][x] === 1) touchesBlack = true;
    if (y - 1 >= 0 && board[y - 1][x] === -1) touchesWhite = true;
  }

  if (touchesBlack && !touchesWhite) return 1;
  if (touchesWhite && !touchesBlack) return -1;
  return 0; // Dane (neutral)
}

/**
 * 估计路径：flood fill 兜底 + probabilityMap 补官子。对应后端 compute_estimated_territory。
 * 先算能确定的封口领地；对 flood fill 留下的中性/单官空点，再用概率图判定：
 * prob > 0.2 → 黑地；prob < -0.2 → 白地。
 */
export function calculateEstimatedTerritory(
  signMap: number[][],
  probabilityMap: number[][],
  deadStones: Set<string>
): TerritoryResult {
  const height = signMap.length;
  const width = signMap[0]?.length || 0;

  const flood = calculateTerritory(signMap, deadStones);
  const territories = flood.territories;
  let blackTerritory = flood.blackTerritory;
  let whiteTerritory = flood.whiteTerritory;

  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      if (signMap[y][x] !== 0 || territories[y][x] !== 0) continue;

      const prob = probabilityMap[y]?.[x] ?? 0;
      if (prob > TERRITORY_THRESHOLD) {
        territories[y][x] = 1;
        blackTerritory++;
      } else if (prob < -TERRITORY_THRESHOLD) {
        territories[y][x] = -1;
        whiteTerritory++;
      }
    }
  }

  return { blackTerritory, whiteTerritory, territories };
}

/** 按子色统计死子数（死子集已给定）。对应后端 count_dead_stones。 */
export function countDeadStones(
  signMap: number[][],
  deadStones: Set<string>
): { blackDeadStones: number; whiteDeadStones: number } {
  let blackDeadStones = 0;
  let whiteDeadStones = 0;

  deadStones.forEach(k => {
    const idx = k.indexOf(',');
    const x = Number(k.slice(0, idx));
    const y = Number(k.slice(idx + 1));
    const sign = signMap[y]?.[x];
    if (sign === 1) blackDeadStones++;
    if (sign === -1) whiteDeadStones++;
  });

  return { blackDeadStones, whiteDeadStones };
}

/**
 * 计算最终分数。对应后端 /api/v1/score 的分数公式：
 *   黑方 = 黑地 + 白死子；白方 = 白地 + 黑死子 + komi。
 * capture（提子数）由对局历史提供，拍照识别场景未知，置 0。
 */
export function computeScore(
  signMap: number[][],
  deadStones: Set<string>,
  komi = 0,
  probabilityMap?: number[][]
): {
  blackTerritory: number;
  whiteTerritory: number;
  blackDeadStones: number;
  whiteDeadStones: number;
  blackScore: number;
  whiteScore: number;
  territoryMap: number[][];
} {
  const counts = countDeadStones(signMap, deadStones);
  const territory = probabilityMap
    ? calculateEstimatedTerritory(signMap, probabilityMap, deadStones)
    : calculateTerritory(signMap, deadStones);

  return {
    blackTerritory: territory.blackTerritory,
    whiteTerritory: territory.whiteTerritory,
    blackDeadStones: counts.blackDeadStones,
    whiteDeadStones: counts.whiteDeadStones,
    blackScore: territory.blackTerritory + counts.whiteDeadStones,
    whiteScore: territory.whiteTerritory + counts.blackDeadStones + komi,
    territoryMap: territory.territories,
  };
}

/**
 * 由后端 /api/v1/deadstones 响应里的死子坐标列表构造 Set<string>。
 * 入参 [{x,y}, ...]，坐标约定与后端一致（x=列, y=行）。
 */
export function deadStonesToSet(deadStones: Array<{ x: number; y: number }>): Set<string> {
  return new Set(deadStones.map(d => key(d.x, d.y)));
}
